### PR TITLE
Jenkins: Present most recent interesting build info

### PR DESCRIPTION
**Pull request by gerhardol**
## Proposed changes
 
Jenkins can start many builds for a commit, for instance for the branch and the pull request.
In some occasions the subsequent build(s) are aborted immediately, getting status NOT_BUILT
It is of course possible to abort builds too

The Jenkins build adapter presents the status for the latest build started for a commit regardless if it is complete or not.
This tries to present the most interesting status instead. 

- Cleanup from Re#, some manual adjustments
- The intervall for retrieving running builds was changed to two minutes in 3fcb4788b42d9b485880eb756e4bdc0af47d321b, but the full day obervable was kept at one minute

## Test methodology

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs [hunk 1]
```diff
@@ -115,21 +115,21 @@ private void AddGetBuildUrl(string projectUrl)
...
-        public class ResponseInfo
+        private class ResponseInfo
         {
             public string Url { get; set; }
             public long Timestamp { get; set; }
             public IEnumerable<JToken> JobDescription { get; set; }
         }
 
-        public class JenkinsCacheInfo
+        private class JenkinsCacheInfo
```
Comment (RussKie) [inline Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs:125]: Do we need the class at all?

File: Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs [hunk 2]
```diff
@@ -161,11 +161,13 @@ private async Task<ResponseInfo> GetBuildInfoTaskAsync(string projectUrl, bool f
                     {
                         try
                         {
-                            if (j["lastBuild"] != null && j["lastBuild"]["timestamp"] != null)
+                            if (j["lastBuild"] == null || j["lastBuild"]["timestamp"] == null)
```
Comment (RussKie) [inline Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs:164]:
```suggestion
                            if (j["lastBuild"] is null || j["lastBuild"]["timestamp"] is null)
```

**Review by RussKie — APPROVED** [review]

File: Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs (hunk 1 above)
Comment (gerhardol) [reply Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs:125]: Yes, it stores the latest known build time for a project url. The adapter first asks for last build time, then requests the full data with all builds if there is anything new (that is more to save the server than the processing in the plugin).

File: Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs (hunk 2 above)
Comment (gerhardol) [reply Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs:164]: Do you want to use the obscure c# 8 "is { }" too?
Forcepushing is null for now

Comment (RussKie) [reply Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs]: In C# 9.0 there's "not is null", we can wait for it :)

**Review by RussKie — COMMENTED** [review]
LGTM

**Comment by RussKie** [comment]
```
JenkinsAdapter.cs(510,13): error SA1508: A closing brace should not be preceded by a blank line 
```

File: Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs (hunk 1 above)
Comment (gerhardol) [reply Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs:125]: But you maybe meant why use a class, why not long directly?
I will change to that
Making a few other touchups, want to test this on Monday/Tuesday before pshing and initiates a merge

**Comment by gerhardol** [comment]
It takes time to test when the dev machine is accessible at night and test by day.
That also caused some more tweaks...
Squashed the first commits, the new changes separately, plan to squash that too in a couple of days (not too hard to review).

File: Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs [hunk 3]
```diff
@@ -320,6 +324,36 @@ private void ObserveBuilds(DateTime? sinceDate, bool? running, IObserver<BuildIn
...
+                    || (!IsBuildCompleted(builds[commit])
+
+                        // Existing status is lower than Completed
+                        && (IsBuildCompleted(newStatus)
+
+                           || (builds[commit] != BuildInfo.BuildStatus.InProgress
+
+                               // Existing status is lower than InProgress
+                               && newStatus != BuildInfo.BuildStatus.Unknown
+                                   && newStatus != BuildInfo.BuildStatus.Stopped)));
```
Comment (RussKie) [inline Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs:348]: I can tell you I'm definitely not a fan of this. It is almost impossible to see conditions boundaries. Can we restructure this with the use of temp variables?

Comment (gerhardol) [reply Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs]: I rewrote using serial if statements instead

**Review by RussKie — COMMENTED** [review]
👍

**Comment by gerhardol** [comment]
@msftbot merge in 24 hours

**Pull request merged**